miden-lib/asm/kernels/transaction/api.masm

```diff
@@ -863,20 +954,99 @@ export.get_note_serial_number
 end
 
 #! Tells the transaction kernel that we are about to execute a procedure on a foreign account.
-#!
-#! Inputs:  [KERNEL_PROCEDURE_HASH, foreign_account_id, pad(11)]
-#! Outputs: [pad(16)]
+#! Loads the data of the specified foreign account to the memory and moves the pointer of the 
+#! currently accessing account to it. 
+#!
+#! Inputs:
+#!   Operand stack: [KERNEL_PROCEDURE_HASH, foreign_account_id, pad(11)]
+#!   Advice map: {
+#!     FOREIGN_ACCOUNT_ID: [[foreign_account_id, 0, 0, account_nonce], VAULT_ROOT, STORAGE_ROOT, CODE_ROOT],
+#!     STORAGE_ROOT: [[STORAGE_SLOT_DATA]],
+#!     CODE_ROOT: [[ACCOUNT_PROCEDURE_DATA]]
+#!   }
+#! Outputs:
+#!   Operand stack: []
 #!
 #! Where:
-#! - foreign_account_id is the ID of the foreign account whose procedure is going to be executed. 
+#! - foreign_account_id is the ID of the foreign account whose procedure is going to be executed.
+#! - FOREIGN_ACCOUNT_ID is the word constructed from the foreign_account_id as follows:
+#!   [foreign_account_id, 0, 0, 0].
+#! - account_nonce is the nonce of the foreign account.
+#! - VAULT_ROOT is the commitment of the foreign account's vault.
+#! - STORAGE_ROOT is the commitment of the foreign account's storage.
+#! - STORAGE_SLOT_DATA is the data contained in the storage slot which is constructed as follows: 
+#!   [SLOT_VALUE, slot_type, 0, 0, 0]
+#! - CODE_ROOT is the commitment of the foreign account's code.
+#! - ACCOUNT_PROCEDURE_DATA is the information about account procedure which is constructed as 
+#!   follows: [PROCEDURE_MAST_ROOT, storage_offset, 0, 0, 0]
 #!
 #! Panics if:
 #! - the current context is not a native context.
 export.start_foreign_context
     # check that this procedure was executed against the native account
     exec.memory::assert_native_account
+    # OS => [KERNEL_PROCEDURE_HASH, foreign_account_id]
 
-    dropw drop # unimplemented!
+    # drop the procedure's hash
+    dropw
+    # OS => [foreign_account_id]
+
+    # get the memory address and a flag whether this account was already loaded.
+    exec.memory::get_foreign_account_ptr
+    # OS => [was_loaded, ptr, foreign_account_id]
+
+    if.true
+        exec.memory::set_current_account_data_ptr drop
+        # OS => []
+    else
+        exec.memory::set_current_account_data_ptr
+        # OS => [foreign_account_id]
+
+        # construct the word with account ID to load the core account data from the advice map
+        push.0.0.0
+        # OS => [0, 0, 0, foreign_account_id]
+
+        # move the core account data to the advice stack
+        adv.push_mapval
+        # OS => [0, 0, 0, foreign_account_id]
+        # AS => [[foreign_account_id, 0, 0, account_nonce], VAULT_ROOT, STORAGE_ROOT, CODE_ROOT]
```

Comment:
One thing that is missing from the code below is that we need to make sure that foreign account state is valid. This can be done by verifying that account hash is present in the account database (i.e., we verify that account hash is in the Merkle tree defined by `ACCT_DB_ROOT_PTR` from the block data section).

Reply:
The Merkle Tree representing the root database is stored in the advice map, right? I can't find the code which saves this Merkle Tree anywhere.

Reply:
The Merkle tree data goes into the Merkle store part of the advice provider. It work similarly to how to load note Merkle trees for verifying note inclusion.

We don't currently load anything related to the account Merkle tree into the advice provider - so, this will be new code.